### PR TITLE
[warmup] Add Support for Warmup 4iE Device [WIP]

### PR DESCRIPTION
**Pull request by jamesmelville**
This binding integrates the Warmup 4iE Thermostat https://www.warmup.co.uk/thermostats/smart/4ie-underfloor-heating, via the API at https://my.warmup.com/.

I logged an issue for this at #7996 .

Per the readme, it currently supports a Bridge to the API, holding the API credentials, and Rooms, which report current temperature and setpoint, and allow setting of an override of the setpoint for a specified duration.

I have tested this successfully on my own installation.

There is a suggested todo list in the readme, but I don't regard any of the suggested functionality as essential and hope this could be merged as-is.

### CODE REVIEW

File: bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:44]: You used the Id in the discovery service. The serial number might be more reasonable.

File: bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:74]: Is there a reason why you didn't document these Channels?

File: bundles/org.openhab.binding.warmup/README.md
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/README.md]: done

File: bundles/org.openhab.binding.warmup/README.md
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/README.md]: removed - is there anywhere you'd recommend to keep this? They're more ideas for if other people want to put effort into extending the binding than a todo list for me. I've focused on the currently implemented features as a complete usable MVP rather than a full client.

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java]: I've reworked the error handling throughout which I hope resolves all the concerns you've highlighted, any feedback will be much appreciated!

File: bundles/org.openhab.binding.warmup/README.md
Comment (fwolter) [reply bundles/org.openhab.binding.warmup/README.md]: You could open a GitHub issue in this repo.

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupConfigurationDTO.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupConfigurationDTO.java]: I'd initially put this in so that the existing session token was not discarded if the username/pw in the configuration were unchanged when the bridge was reconfigured. I've now taken it out because it's probably unnecessary complexity and I've not seen the API get unhappy with frequent logins.

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java]: Resolved - thank you for correcting my misunderstanding of this method.

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java]: Resolved

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java]: Configuration DTOs are now used throughout

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java]: I removed this - the status is now set once a first attempt to call API is completed

File: bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:17]: added

File: bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/binding/binding.xml
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/binding/binding.xml]: I extended the description, I hope it's now suitable

File: bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:47]: Thanks - I think this is all aligned now

File: bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/resources/ESH-INF/thing/thing-types.xml:77]: No reason, I've added them to the updated readme

**Comment by jamesmelville** [comment]
Hi Fabian @fwolter, thanks for the review. I've updated everything you requested, fixed the checkstyle warnings, and my build is now passing. I've only had one owner of a device provide feedback https://community.openhab.org/t/warmup-thermostat-41e/37394/5, but I'll see if I can get any more.

File: bundles/org.openhab.binding.warmup/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/README.md:52]: Can you mention these modes?

File: bundles/org.openhab.binding.warmup/README.md
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/README.md:53]: Do you mean the mode in `runMode` with "previously configured mode"?

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java:104]: Bindings should only log to error if something severe happened, like the detection of a bug in your code. This can be warn.

Also, it's good practice to log exceptions where they are catched. This prevents multiple logging of the same exception in different locations.

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:55]: Better validate the value of `config.refreshIntervalSec` as it can be 0 if the user doesn't configure it. Then, an IllegalArgumentException would be raised.

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java:47]: The configuration should be read in `initialize()`, since the handler object is re-used when the configuration is updated. Then, `dispose()` followed by `initialize()` will be invoked, without calling the constructor.

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java:91]: You could make `DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")` a static final constant to save resources.

Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java]: Code removed as it was unused

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java]: Thank you, this is resolved, although I hope the @SuppressWarnings("null") I added is ok? https://github.com/openhab/openhab-addons/pull/7998/commits/7bc9ed18881442a51c73f95fe123773f6d04e482#diff-782a12acc9c698d8cf1b524ad394758e

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java]: Implemented guard against value being below 10 specified in thing.xml https://github.com/openhab/openhab-addons/pull/7998/commits/7bc9ed18881442a51c73f95fe123773f6d04e482#diff-028f0585e493ccac6de2540a9f878ba5

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/api/MyWarmupApi.java]: Thanks, updated

File: bundles/org.openhab.binding.warmup/README.md
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/README.md]: I've added further explanation (based on my limited knowledge!).

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java
Comment (fwolter) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java]: Sorry for the delay. Please do a null check instead of suppressing the warning. I'm going to approve the code after the fix. @Hilbrand you could do a second review if you want.

File: bundles/org.openhab.binding.warmup/README.md
Comment (Hilbrand) [inline bundles/org.openhab.binding.warmup/README.md:7]: Can you put each sentence on it's own line, that makes viewing (future) diffs more easily.
```suggestion
This API is not known to be documented publicly.
The binding api implementation has been derived from the implementations at https://github.com/alyc100/SmartThingsPublic/blob/master/devicetypes/alyc100/warmup-4ie.src/warmup-4ie.groovy and https://github.com/alex-0103/warmup4IE/blob/master/warmup4ie/warmup4ie.py, and enhanced by inspecting the GraphQL endpoint.
```

File: bundles/org.openhab.binding.warmup/README.md
Comment (Hilbrand) [inline bundles/org.openhab.binding.warmup/README.md:34]: Can you format the tables. You can use http://markdowntable.com/

File: bundles/org.openhab.binding.warmup/pom.xml
Comment (Hilbrand) [inline bundles/org.openhab.binding.warmup/pom.xml:10]: Can you rebase to 2.5.9. With `git pull --rebase` and than `git push --force-with-lease` to get the changes back to GitHub
```suggestion
    <version>2.5.9-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupHandlerFactory.java
Comment (Hilbrand) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupHandlerFactory.java:68]: Can you implement discovery registration as `ThingHandlerService` See this preview documentation about how that looks like: https://deploy-preview-1262--openhab-docs-preview.netlify.app/docs/developer/bindings/#discovery-that-is-bound-to-a-thing

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java
Comment (Hilbrand) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/discovery/WarmupDiscoveryService.java:73]: In general older results are removed by adding the follow line to the `stopScan()` method:
```
removeOlderResults(getTimestampOfLastScan());
```

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java
Comment (Hilbrand) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/MyWarmupAccountHandler.java:127]: A discovery should only be triggered from the discovery service itself. So this part might not be a good idea. What is the reason to call it here?

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java
Comment (Hilbrand) [inline bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupThingHandler.java:45]: You should probably set the status of the thing to `UNKNOWN` here, or is the status handled elsewhere?

**Review by Hilbrand — CHANGES_REQUESTED** [review]
Thank you for your contribution. Just a few comments, mostly related to discovery.

Make sure you run `mvn spotless:apply` on you binding before pushing the code to GitHub.

If you're able to process these review comments in the next days we can add this new binding to 2.5.9, otherwise you'll have to migrate it to openHAB 3.0.

File: bundles/org.openhab.binding.warmup/pom.xml
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/pom.xml:10]: Migrated to v3.0.0

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/RoomHandler.java]: resolved in https://github.com/openhab/openhab-addons/pull/8562/commits/7be6f8d8ab878b0c37f5040555ce9e4f9707188c

File: bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupHandlerFactory.java
Comment (jamesmelville) [reply bundles/org.openhab.binding.warmup/src/main/java/org/openhab/binding/warmup/internal/handler/WarmupHandlerFactory.java:68]: Resolved in https://github.com/openhab/openhab-addons/pull/8562/commits/4c41f60f5463036c3e0300170cf496166db005c7

**Comment by jamesmelville** [comment]
Hi @Hilbrand @fwolter, thank you for your reviews. Unfortunately I've been on holiday so missed the deadline for 2.5.9. I've made all the changes you've requested along with updating the pull request for 3.0.0 here #8562.

**Comment by fwolter** [comment]
Closing in favor of #8562

**Pull request closed without merging**
